### PR TITLE
Remove support for Python 2 (#76) + remove deprecated commands (#26)

### PR DESCRIPTION
**Pull request by PawelLipski**

### CODE REVIEW

**Review by ilai-deutel — COMMENTED** [review]
Looks good overall! A few comments about the transition to Python 3:

- [`shlex.quote`](https://docs.python.org/3/library/shlex.html#shlex.quote) (new in Python 3.3) could be used to replace `shell_escape`
- [`subprocess.run`](https://docs.python.org/3/library/subprocess.html#subprocess.run) is the recommended approach to invoking subprocesses since Python 3.5
- `# -*- coding: utf-8 -*-` is not needed anymore since UTF-8 is the default for Python 3
- In `back_up_definition_file`, I think the `definition_file` is opened and never closed

File: git_machete/cmd.py [hunk 1]
```diff
@@ -2489,11 +2465,11 @@ def version():
...
 
-def main():
-    launch(sys.argv[1:])
-
-
 def launch(orig_args):
+    if sys.version_info[0] == 2:
+        sys.stderr.write("Python 2.x is no longer supported. Please switch to Python 3.\n")
+        sys.exit(1)
+
```
[the remark this answers is not in the document]
Comment (PawelLipski) [reply git_machete/cmd.py:2981]: Done, added. Actually, do you think it's worthwhile to provide the support for Python versions 3.0-3.4, or are they rare enough nowadays that just support for >=3.5 is sufficient?

File: tox.ini [hunk 2]
```diff
@@ -1,5 +1,5 @@
 [tox]
-envlist = pep8,py27,py36,py38
+envlist = pep8,py37
```
[the remark this answers is not in the document]
Comment (PawelLipski) [reply tox.ini]: Done, added

**Comment by PawelLipski** [comment]
> Looks good overall! A few comments about the transition to Python 3:
Thanks for the review!

> * [`shlex.quote`](https://docs.python.org/3/library/shlex.html#shlex.quote) (new in Python 3.3) could be used to replace `shell_escape`

Actually, `shlex.quote` doesn't replace `\t` with `$'\t'`, just keeps a literal` \x07` byte unchanged  - which doesn't work with copy-paste from terminal well. Sticking to the current implementation.

> * [`subprocess.run`](https://docs.python.org/3/library/subprocess.html#subprocess.run) is the recommended approach to invoking subprocesses since Python 3.5

Done, changed.

> * `# -*- coding: utf-8 -*-` is not needed anymore since UTF-8 is the default for Python 3

Done, removed.

> * In `back_up_definition_file`, I think the `definition_file` is opened and never closed

Hmm but definition_file is opened within `with` statement... so it should be closed in context manager's `__exit__()`.

File: git_machete/cmd.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (ilai-deutel) [reply git_machete/cmd.py:2981]: I think the main issue with supporting 3.0-3.4 is that it prevents you from using [PEP 484](https://www.python.org/dev/peps/pep-0484/) type hints.
These versions have all reached their end-of-life (see [the devguide](https://devguide.python.org/devcycle/#end-of-life-branches)), but I'm not sure if they are still widely used (3.4 was released 6 years ago).

**Comment by ilai-deutel** [comment]
> > * In `back_up_definition_file`, I think the `definition_file` is opened and never closed
> 
> Hmm but definition_file is opened within `with` statement... so it should be closed in context manager's `__exit__()`.

I think `definition_file + "~"` is, but not `definition_file`:

https://github.com/VirtusLab/git-machete/blob/b44a34154c30c17ab851b6d76ccce0b4fe417cdd/git_machete/cmd.py#L273

File: git_machete/cmd.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (PawelLipski) [reply git_machete/cmd.py:2981]: Ok, let's require >=3.5 then

**Comment by PawelLipski** [comment]
> > > * In `back_up_definition_file`, I think the `definition_file` is opened and never closed
> > 
> > 
> > Hmm but definition_file is opened within `with` statement... so it should be closed in context manager's `__exit__()`.
> 
> I think `definition_file + "~"` is, but not `definition_file`:
> 
> https://github.com/VirtusLab/git-machete/blob/b44a34154c30c17ab851b6d76ccce0b4fe417cdd/git_machete/cmd.py#L273

Whoops right, good spot ;D I'll fix that on another patch-level release (v2.13.2), using `shutil.copyfile` instead

**Comment by PawelLipski** [comment]
Work will be continued on @mkondratek's PR #107

**Pull request closed without merging**